src/libCrypto/EthCrypto.cpp

```diff
@@ -0,0 +1,183 @@
+#include <openssl/obj_mac.h>  // for NID_secp192k1
+#include <openssl/sha.h>      //for SHA512_DIGEST_LENGTH
+#include <ethash/keccak.hpp>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <string>
+
+using namespace std;
+
+// Inspiration from:
+// https://stackoverflow.com/questions/10906524
+// https://stackoverflow.com/questions/57385412/
+// https://medium.com/mycrypto/the-magic-of-digital-signatures-on-ethereum-98fe184dc9c7
+
+// Prefix signed txs in Ethereum with Keccak256("\x19Ethereum Signed
+// Message:\n32" + Keccak256(message))
+constexpr uint8_t prelude[] = {25,  69,  116, 104, 101, 114, 101, 117, 109,
+                               32,  83,  105, 103, 110, 101, 100, 32,  77,
+                               101, 115, 115, 97,  103, 101, 58,  10,  48};
+
+// https://stackoverflow.com/questions/57385412/
+void SetOpensslSignature(const std::string& sSignatureInHex, ECDSA_SIG* pSign) {
+  // Openssl uses raw pointers and macros for freeing but can force it into
+  // smart pointers methodology
+  std::unique_ptr<BIGNUM, std::function<void(BIGNUM*)>> rr(
+      NULL, [](BIGNUM* b) { BN_free(b); });
+  BIGNUM* r_ptr = rr.get();
+  std::unique_ptr<BIGNUM, std::function<void(BIGNUM*)>> ss(
+      NULL, [](BIGNUM* b) { BN_free(b); });
+  BIGNUM* s_ptr = ss.get();
+
+  std::string sSignatureR =
+      sSignatureInHex.substr(0, sSignatureInHex.size() / 2);
+  std::string sSignatureS = sSignatureInHex.substr(sSignatureInHex.size() / 2);
+
+  BN_hex2bn(&r_ptr, sSignatureR.c_str());
+  BN_hex2bn(&s_ptr, sSignatureS.c_str());
+
+  ECDSA_SIG_set0(pSign, r_ptr, s_ptr);
+}
+
+// This function needs to set the key but also needs to decomprss the
+// coordinates using EC_POINT_set_compressed_coordinates_GFp sPubKeyString
+// should be hex \0 terminated string with no compression bit on it
+bool SetOpensslPublicKey(const char* sPubKeyString, EC_KEY* pKey) {
+  // X co-ordinate
+  std::unique_ptr<BIGNUM, std::function<void(BIGNUM*)>> gx(
+      NULL, [](BIGNUM* b) { BN_free(b); });
+
+  BIGNUM* gx_ptr = gx.get();
+
+  EC_KEY_set_asn1_flag(pKey, OPENSSL_EC_NAMED_CURVE);
+
+  // From
+  // https://www.oreilly.com/library/view/mastering-ethereum/9781491971932/ch04.html
+  // The first byte indicates whether the y coordinate is odd or even
+  int y_chooser_bit = 0;
+
+  if (sPubKeyString[1] == '2') {
+    y_chooser_bit = 0;
+  } else if (sPubKeyString[1] == '3') {
+    y_chooser_bit = 1;
+  } else {
+    std::cout << "Received badly set signature bit! Should be 2 or 3 and got: "
+              << sPubKeyString[1] << std::endl;
+  }
+
+  // Don't want the first byte
+  if (!BN_hex2bn(&gx_ptr, sPubKeyString + 2)) {
+    std::cout << "***** Error getting to x binary format" << std::endl;
+  }
+
+  // Create a new curve group
+  // Refactor: probably can have static/constexpr curve, need to check
+  std::unique_ptr<EC_GROUP, std::function<void(EC_GROUP*)>> curve_group(
```

Comment:
Ok maybe worth having it here, but maybe a `using` directrive or a typedef for this somewhere, so that you don't have to repeat the lambda. Or a few unique_ptr specializations etc.

Reply:
sounds great